HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc

```diff
@@ -0,0 +1,343 @@
+  desc.add<int>("MinNmuonHitsL3", 0);
+  desc.add<int>("MinNhitsL3", 0);
+  desc.add<double>("MaxEtaL2", 2.5);
+  desc.add<double>("MinPtL2", 3.0);
+  desc.add<double>("MaxNormalizedChi2L2", 9999.0);
+  desc.add<int>("MinNmuonHitsL2", 0);
+  desc.add<std::vector<double> >("AbsEtaBinsL2", std::vector<double>(1, 9999.));
+  desc.add<std::vector<int> >("MinNstationsL2", std::vector<int>(1, 1));
+  desc.add<std::vector<int> >("MinNhitsL2", std::vector<int>(1, 0));
+  desc.add<std::vector<int> >("MinNchambersL2", std::vector<int>(1, 0));
+  descriptions.add("hltMuonL3orL2PreFilter", desc);
+}
+
+// ------------ method called to produce the data  ------------
+bool HLTMuonL3orL2PreFilter::hltFilter(edm::Event& iEvent,
+                                       const edm::EventSetup& iSetup,
+                                       trigger::TriggerFilterObjectWithRefs& filterproduct) const {
+  // All HLT filters must create and fill an HLT filter object,
+  // recording any reconstructed physics objects satisfying (or not)
+  // this HLT filter, and place it in the Event.
+
+  if (saveTags())
+    filterproduct.addCollectionTag(candTag_);
+
+  // get hold of trks
+  Handle<RecoChargedCandidateCollection> mucands;
+  iEvent.getByToken(candToken_, mucands);
+  Handle<TriggerFilterObjectWithRefs> previousLevelCands;
+  iEvent.getByToken(previousCandToken_, previousLevelCands);
+  Handle<RecoChargedCandidateCollection> L2mucands;
+  iEvent.getByToken(L2CandToken_, L2mucands);
+  vector<RecoChargedCandidateRef> vcands;
+  if (previousLevelCands.isValid()) {
+    previousLevelCands->getObjects(TriggerMuon, vcands);
+  }
+
+  Handle<BeamSpot> recoBeamSpotHandle;
+  iEvent.getByToken(beamspotToken_, recoBeamSpotHandle);
+
+  // Number of objects passing the L3 Trigger or L2:
+  int n = 0;
+  bool firstIteration = true;
+  for (unsigned int iMuL2 = 0; iMuL2 < L2mucands->size(); iMuL2++) {
+    RecoChargedCandidateRef L2cand(L2mucands, iMuL2);
+    LogDebug("HLTMuonL3orL2PreFilter") << "cand isNonnull " << L2cand.isNonnull();
+
+    //    //did this candidate triggered at previous stage.
+    //    if (matchPreviousCand_ && !triggerdByPreviousLevel(cand, vcands))
+    //      continue;
+    bool FailedL2 = false;
+    if (std::abs(L2cand->eta()) > max_Eta_L2_)
+      FailedL2 = true;
+
+    TrackRef L2tk = L2cand->track();
+    LogDebug("HLTMuonL3orL2PreFilter") << " Muon in loop, q*pt= " << L2tk->charge() * L2tk->pt() << " ("
+                                       << L2cand->charge() * L2cand->pt() << ") "
+                                       << ", eta= " << L2tk->eta() << " (" << L2cand->eta() << ") "
+                                       << ", hits= " << L2tk->numberOfValidHits() << ", d0= " << L2tk->d0()
+                                       << ", dz= " << L2tk->dz();
+
+    //normalizedChi2 cut
+    if (L2tk->normalizedChi2() > max_NormalizedChi2_L2_)
+      FailedL2 = true;
+
+    if (recoBeamSpotHandle.isValid()) {
```

Comment:
```suggestion
    if (useBeamspot_) {
      if (not recoBeamSpotHandle.isValid()) { throw cms::Exception("") << ""; }
```
(the suggestion is pseudo-code..)

If the beamspot handle is invalid (e.g. due to a typo in the InputTag), certain cuts/selections are silently skipped. It would be safer to add a boolean `useBeamspot` in the PSet, and to throw an exception if `useBeamspot==true and recoBeamSpotHandle.isValid() == false`.

Reply:
I have added the exception but I think that it is not necessary to include another parameter to use the beam spot. If you want to use the cuts that are implemented under this bracket you can change for example the parameter max_Dz_. The beam spot is used only to compute Dz (the impact parameter in the z direction).